BareResponseImpl throws ISE in Subscriber.onNext method

## Environment Details
* Helidon Version: 2.1.0

And spec rules [§ 2.13](https://github.com/reactive-streams/reactive-streams-jvm#2.13) and [§ 2.8](https://github.com/reactive-streams/reactive-streams-jvm#2.8) are violated.

```java
    @Override
    public void onNext(DataChunk data) {
        if (internallyClosed.get()) {
            throw new IllegalStateException("Response is already closed!");
        }
```

Expected behavior should be for BareResponseImpl to send cancel signal upstream and in case of subsequent onNext signal log a message about dropping DataChunk.

Current behavior causes problems for example with SSE, when multiple sse clients connect and one of them drops, ISE is thrown and app output get quickly polluted even in such legal situation.

## Reproducer 
```java
    SubmissionPublisher<String> broadCastPublisher = new SubmissionPublisher<>();

    void init(@Observes @Initialized(ApplicationScoped.class) Object o) {
        Multi.interval(300, TimeUnit.MILLISECONDS, Executors.newSingleThreadScheduledExecutor())
                .forEach(i -> broadCastPublisher.submit("message" + i));
    }

    @GET
    @Path("sse-reproduce")
    @Produces(MediaType.SERVER_SENT_EVENTS)
    public void listenToEvents(@Context Flow.Subscriber<String> sink) {
        broadCastPublisher.subscribe(sink);
    }
```
Curl endpoind, reveive few messages and then kill connection with Ctrl+C:
```sh
curl localhost:8080/example/sse-reproduce
data: message335
data: message336
data: message337
data: message338
data: message339
^C
```

```
2020.11.03 18:56:11 SEVERE org.glassfish.jersey.media.sse.internal.JerseyEventSink Thread[ForkJoinPool.commonPool-worker-13,5,main]: Processing onNext signal failed.
java.io.IOException: java.lang.IllegalStateException: java.lang.IllegalStateException: Response is already closed!
	at io.helidon.common.reactive.MultiFromOutputStream.doPublish(MultiFromOutputStream.java:180)
	at io.helidon.common.reactive.MultiFromOutputStream.publish(MultiFromOutputStream.java:158)
	at io.helidon.common.reactive.MultiFromOutputStream.write(MultiFromOutputStream.java:96)
	at io.helidon.webserver.jersey.ResponseWriter$1.write(ResponseWriter.java:52)
	at org.glassfish.jersey.message.internal.CommittingOutputStream.write(CommittingOutputStream.java:185)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$UnCloseableOutputStream.write(WriterInterceptorExecutor.java:271)
	at org.glassfish.jersey.media.sse.OutboundEventWriter$1.write(OutboundEventWriter.java:127)
	at java.base/java.io.OutputStream.write(OutputStream.java:157)
	at java.base/sun.nio.cs.StreamEncoder.writeBytes(StreamEncoder.java:242)
	at java.base/sun.nio.cs.StreamEncoder.implFlushBuffer(StreamEncoder.java:321)
	at java.base/sun.nio.cs.StreamEncoder.implFlush(StreamEncoder.java:325)
	at java.base/sun.nio.cs.StreamEncoder.flush(StreamEncoder.java:159)
	at java.base/java.io.OutputStreamWriter.flush(OutputStreamWriter.java:254)
	at org.glassfish.jersey.message.internal.ReaderWriter.writeToAsString(ReaderWriter.java:170)
	at org.glassfish.jersey.message.internal.AbstractMessageReaderWriterProvider.writeToAsString(AbstractMessageReaderWriterProvider.java:107)
	at org.glassfish.jersey.message.internal.StringMessageProvider.writeTo(StringMessageProvider.java:76)
	at org.glassfish.jersey.message.internal.StringMessageProvider.writeTo(StringMessageProvider.java:36)
	at org.glassfish.jersey.media.sse.OutboundEventWriter.writeTo(OutboundEventWriter.java:113)
	at org.glassfish.jersey.media.sse.OutboundEventWriter.writeTo(OutboundEventWriter.java:43)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.invokeWriteTo(WriterInterceptorExecutor.java:242)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor$TerminalWriterInterceptor.aroundWriteTo(WriterInterceptorExecutor.java:227)
	at org.glassfish.jersey.message.internal.WriterInterceptorExecutor.proceed(WriterInterceptorExecutor.java:139)
	at org.glassfish.jersey.message.internal.MessageBodyFactory.writeTo(MessageBodyFactory.java:1116)
	at org.glassfish.jersey.server.ChunkedOutput$1.call(ChunkedOutput.java:231)
	at org.glassfish.jersey.server.ChunkedOutput$1.call(ChunkedOutput.java:198)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292)
	at org.glassfish.jersey.internal.Errors.process(Errors.java:219)
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:314)
	at org.glassfish.jersey.server.ChunkedOutput.flushQueue(ChunkedOutput.java:198)
	at org.glassfish.jersey.server.ChunkedOutput.write(ChunkedOutput.java:184)
	at org.glassfish.jersey.media.sse.internal.JerseyEventSink.onNext(JerseyEventSink.java:93)
	at java.base/java.util.concurrent.SubmissionPublisher$BufferedSubscription.consumeNext(SubmissionPublisher.java:1307)
	at java.base/java.util.concurrent.SubmissionPublisher$BufferedSubscription.takeItems(SubmissionPublisher.java:1296)
	at java.base/java.util.concurrent.SubmissionPublisher$BufferedSubscription.consume(SubmissionPublisher.java:1253)
	at java.base/java.util.concurrent.SubmissionPublisher$ConsumerTask.exec(SubmissionPublisher.java:930)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1016)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1665)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1598)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
Caused by: java.lang.IllegalStateException: java.lang.IllegalStateException: Response is already closed!
	at io.helidon.common.reactive.EmittingPublisher.internalEmit(EmittingPublisher.java:318)
	at io.helidon.common.reactive.EmittingPublisher$State$4.emit(EmittingPublisher.java:360)
	at io.helidon.common.reactive.EmittingPublisher.emit(EmittingPublisher.java:198)
	at io.helidon.common.reactive.MultiFromOutputStream.doPublish(MultiFromOutputStream.java:174)
	... 39 more
Caused by: java.lang.IllegalStateException: Response is already closed!
	at io.helidon.webserver.BareResponseImpl.onNext(BareResponseImpl.java:310)
	at io.helidon.webserver.BareResponseImpl.onNext(BareResponseImpl.java:56)
	at io.helidon.media.common.MessageBodyContext$EventingSubscriber.onNext(MessageBodyContext.java:334)
	at io.helidon.media.common.MessageBodyContext$EventingSubscriber.onNext(MessageBodyContext.java:300)
	at io.helidon.common.reactive.MultiMapperPublisher$MapperSubscriber.onNext(MultiMapperPublisher.java:80)
	at io.helidon.common.reactive.EmittingPublisher.internalEmit(EmittingPublisher.java:310)
	... 42 more
```

--- a/webserver/webserver/pom.xml
+++ b/webserver/webserver/pom.xml
@@ -79,6 +79,30 @@
                     </execution>
                 </executions>
             </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <executions>
+                    <execution>
+                        <goals>
+                            <goal>integration-test</goal>
+                            <goal>verify</goal>
+                        </goals>
+                    </execution>
+                </executions>
+                <configuration>
+                    <includes>
+                        <include>**/*TckTest.java</include>
+                    </includes>
+                </configuration>
+                <dependencies>
+                    <dependency>
+                        <groupId>org.apache.maven.surefire</groupId>
+                        <artifactId>surefire-testng</artifactId>
+                        <version>${version.lib.surefire.testng}</version>
+                    </dependency>
+                </dependencies>
+            </plugin>
         </plugins>
     </build>
 
@@ -193,6 +217,11 @@
         <dependency>
             <groupId>io.helidon.config</groupId>
             <artifactId>helidon-config-testing</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.reactivestreams</groupId>
+            <artifactId>reactive-streams-tck-flow</artifactId>
             <scope>test</scope>
         </dependency>
     </dependencies>

--- a/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
@@ -222,7 +222,9 @@ class BareResponseImpl implements BareResponse {
             responseFuture.complete(this);
         } else {
             LOGGER.finer(() -> log("Response completion failed %s", throwable));
-            Optional.ofNullable(subscription).ifPresent(Flow.Subscription::cancel);
+            if (subscription != null) {
+                subscription.cancel();
+            }
             internallyClosed.set(true);
             responseFuture.completeExceptionally(throwable);
         }
@@ -236,8 +238,8 @@ class BareResponseImpl implements BareResponse {
      */
     private void completeInternal(Throwable throwable) {
         boolean wasClosed = !internallyClosed.compareAndSet(false, true);
-        if(wasClosed){
-            Optional.ofNullable(subscription).ifPresent(Flow.Subscription::cancel);
+        if (wasClosed && subscription != null) {
+            subscription.cancel();
         }
         orderedWrite(() -> completeInternalPipe(wasClosed, throwable));
     }

--- a/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
@@ -21,7 +21,6 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
-import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Flow;
 import java.util.concurrent.atomic.AtomicBoolean;

--- a/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
+++ b/webserver/webserver/src/main/java/io/helidon/webserver/BareResponseImpl.java
@@ -21,6 +21,7 @@ import java.util.Arrays;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
+import java.util.Optional;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.Flow;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -221,6 +222,7 @@ class BareResponseImpl implements BareResponse {
             responseFuture.complete(this);
         } else {
             LOGGER.finer(() -> log("Response completion failed %s", throwable));
+            Optional.ofNullable(subscription).ifPresent(Flow.Subscription::cancel);
             internallyClosed.set(true);
             responseFuture.completeExceptionally(throwable);
         }
@@ -234,6 +236,9 @@ class BareResponseImpl implements BareResponse {
      */
     private void completeInternal(Throwable throwable) {
         boolean wasClosed = !internallyClosed.compareAndSet(false, true);
+        if(wasClosed){
+            Optional.ofNullable(subscription).ifPresent(Flow.Subscription::cancel);
+        }
         orderedWrite(() -> completeInternalPipe(wasClosed, throwable));
     }
 
@@ -333,35 +338,34 @@ class BareResponseImpl implements BareResponse {
 
     @Override
     public void onSubscribe(Flow.Subscription subscription) {
-        this.subscription = subscription;
+        if (this.subscription != null) {
+            subscription.cancel();
+            return;
+        }
+        this.subscription = Objects.requireNonNull(subscription, "subscription is null");
         subscription.request(1);
     }
 
     @Override
     public void onNext(DataChunk data) {
-        if (internallyClosed.get()) {
+        Objects.requireNonNull(data, "DataChunk is null");
+        if (data.isFlushChunk()) {
+            if (prevRequestChunk == null) {
+                ctx.flush();
+            } else {
+                prevRequestChunk = prevRequestChunk.thenRun(ctx::flush);
+            }
             subscription.request(1);
-            throw new IllegalStateException("Response is already closed!");
+            return;
         }
-        if (data != null) {
-            if (data.isFlushChunk()) {
-                if (prevRequestChunk == null) {
-                   ctx.flush();
-                } else {
-                   prevRequestChunk = prevRequestChunk.thenRun(ctx::flush);
-                }
-                subscription.request(1);
-                return;
-            }
 
-            if (lengthOptimization && firstChunk == null) {
-                firstChunk = data.isReadOnly() ? data : data.duplicate();      // cache first chunk
-                subscription.request(1);
-                return;
-            }
-
-            orderedWrite(() -> onNextPipe(data));
+        if (lengthOptimization && firstChunk == null) {
+            firstChunk = data.isReadOnly() ? data : data.duplicate();      // cache first chunk
+            subscription.request(1);
+            return;
         }
+
+        orderedWrite(() -> onNextPipe(data));
     }
 
     /**
@@ -461,18 +465,13 @@ class BareResponseImpl implements BareResponse {
 
     @Override
     public void onError(Throwable thr) {
+        Objects.requireNonNull(thr, "throwable is null");
         completeInternal(thr);
-        if (subscription != null) {
-            subscription.cancel();
-        }
     }
 
     @Override
     public void onComplete() {
         completeInternal(null);
-        if (subscription != null) {
-            subscription.cancel();
-        }
     }
 
     @Override

--- a/webserver/webserver/src/test/java/io/helidon/webserver/BareResponseSubscriberTckTest.java
+++ b/webserver/webserver/src/test/java/io/helidon/webserver/BareResponseSubscriberTckTest.java
@@ -5,14 +5,13 @@
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
- * http://www.apache.org/licenses/LICENSE-2.0
+ *     http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
- *
  */
 
 package io.helidon.webserver;

--- a/webserver/webserver/src/test/java/io/helidon/webserver/BareResponseSubscriberTckTest.java
+++ b/webserver/webserver/src/test/java/io/helidon/webserver/BareResponseSubscriberTckTest.java
@@ -1,0 +1,98 @@
+/*
+ * Copyright (c) 2021 Oracle and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package io.helidon.webserver;
+
+import io.helidon.common.http.DataChunk;
+import io.netty.channel.Channel;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.handler.codec.http.EmptyHttpHeaders;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpVersion;
+import org.mockito.Mockito;
+import org.reactivestreams.tck.TestEnvironment;
+import org.reactivestreams.tck.flow.FlowSubscriberWhiteboxVerification;
+
+import java.nio.charset.StandardCharsets;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Flow;
+
+public class BareResponseSubscriberTckTest extends FlowSubscriberWhiteboxVerification<DataChunk> {
+
+    protected BareResponseSubscriberTckTest() {
+        super(new TestEnvironment(200));
+    }
+
+    @Override
+    public Flow.Subscriber<DataChunk> createFlowSubscriber(WhiteboxSubscriberProbe<DataChunk> probe) {
+        ChannelHandlerContext ctx = Mockito.mock(ChannelHandlerContext.class);
+        HttpRequest httpRequest = Mockito.mock(HttpRequest.class);
+        Channel channel = Mockito.mock(Channel.class);
+        ChannelFuture channelFuture = Mockito.mock(ChannelFuture.class);
+
+        Mockito.when(httpRequest.headers()).thenReturn(EmptyHttpHeaders.INSTANCE);
+        Mockito.when(httpRequest.protocolVersion()).thenReturn(HttpVersion.HTTP_1_0);
+        Mockito.when(ctx.channel()).thenReturn(channel);
+        Mockito.when(channel.closeFuture()).thenReturn(channelFuture);
+
+        return new BareResponseImpl(ctx,
+                httpRequest,
+                () -> true,
+                CompletableFuture.completedFuture(null),
+                0L) {
+            @Override
+            public void onSubscribe(Flow.Subscription subscription) {
+                super.onSubscribe(subscription);
+                probe.registerOnSubscribe(new SubscriberPuppet() {
+                    @Override
+                    public void triggerRequest(long elements) {
+                        subscription.request(elements);
+                    }
+
+                    @Override
+                    public void signalCancel() {
+                        subscription.cancel();
+                    }
+                });
+            }
+
+            @Override
+            public void onNext(DataChunk data) {
+                super.onNext(data);
+                probe.registerOnNext(data);
+            }
+
+            @Override
+            public void onError(Throwable thr) {
+                super.onError(thr);
+                probe.registerOnError(thr);
+            }
+
+            @Override
+            public void onComplete() {
+                super.onComplete();
+                probe.registerOnComplete();
+            }
+        };
+    }
+
+    @Override
+    public DataChunk createElement(int i) {
+        return DataChunk.create(String.valueOf(i).getBytes(StandardCharsets.UTF_8));
+    }
+}

